TYPO3.Flow/Classes/TYPO3/Flow/Object/Proxy/Compiler.php

```diff
@@ -227,12 +230,12 @@ protected function cacheOriginalClassFileAndProxyCode($className, $pathAndFilena
         $classCode = $this->stripOpeningPhpTag($classCode);
 
         $classNameSuffix = self::ORIGINAL_CLASSNAME_SUFFIX;
-        $classCode = preg_replace_callback('/^([a-z\s]*)(interface|class)\s+([a-zA-Z0-9_]+)/m', function ($matches) use ($pathAndFilename, $classNameSuffix) {
+        $classCode = preg_replace_callback('/^([a-z\s]*?)(final\s+)?(interface|class)\s+([a-zA-Z0-9_]+)/m', function ($matches) use ($pathAndFilename, $classNameSuffix, $proxyClassCode) {
             $classNameAccordingToFileName = basename($pathAndFilename, '.php');
-            if ($matches[3] !== $classNameAccordingToFileName) {
-                throw new Exception('The name of the class "' . $matches[3] . '" is not the same as the filename which is "' . basename($pathAndFilename) . '". Path: ' . $pathAndFilename, 1398356897);
+            if ($matches[4] !== $classNameAccordingToFileName) {
+                throw new Exception('The name of the class "' . $matches[4] . '" is not the same as the filename which is "' . basename($pathAndFilename) . '". Path: ' . $pathAndFilename, 1398356897);
             }
-            return $matches[1] . $matches[2] . ' ' . $matches[3] . $classNameSuffix;
+            return $matches[1] . $matches[3] . ' ' . $matches[4] . $classNameSuffix;
```

Comment:
This removes any `final` modifier from the `*_Original` class (it will be moved to the proxy below)